### PR TITLE
add skeleton for barn prize pool contracts

### PR DESCRIPTION
**Pull request by sstrenev**
* Added initial skeleton for the Barn Bridge Prize Pool contracts with the base functions.

### CODE REVIEW

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  function _supply(uint256) internal override {
+    IERC20Upgradeable assetToken = _token();
+    uint256 total = assetToken.balanceOf(address(this));
+    assetToken.approve(address(barn), total);
+    barn.deposit(total);
+  }
+
+  /// @dev Allows a user to supply asset tokens in exchange for yield-bearing tokens
+  /// to be held in escrow by the Yield Service
+  function _supplySpecific(uint256 amount) internal {
```
Comment (Perseverance) [inline contracts/prize-pool/barnbridge/BarnPrizePool.sol]: I think `yearn` pool used this in order to do some logic around fees. Can you check for me if we need to perform similar fees logic, or remove this method

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  /// @dev Gets the balance of the underlying assets held by the Yield Service
+  /// @return The underlying balance of asset tokens
+  function _balance() internal override returns (uint256) {
+    uint256 balance = barn.balanceOf(address(this));
+    return balance;
+  }
+
+  /// @dev Allows a user to supply asset tokens in exchange for yield-bearing tokens
```
Comment (Perseverance) [inline contracts/prize-pool/barnbridge/BarnPrizePool.sol]: I'd add comment that this method ignores the parameter and always deposits everything

File: contracts/external/barnbridge/BarnInterface.sol [hunk 3]
```diff
@@ -0,0 +1,13 @@
+pragma solidity >=0.6.0 <0.7.0;
+
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
+
+interface BarnInterface is IERC20Upgradeable {
```
Comment (Perseverance) [inline contracts/external/barnbridge/BarnInterface.sol]: Can you point me to the Barn staking contract that you have used in order to find these functions?

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  /// @return True if the token may be awarded, false otherwise
+  function _canAwardExternal(address _externalToken) internal override view returns (bool) {
+    return _externalToken != address(barn) && _externalToken != address(barn.token());
+  }
+
+  /// @dev Allows a user to redeem yield-bearing tokens in exchange for the underlying
+  /// asset tokens held in escrow by the Yield Service
+  /// @param amount The amount of underlying tokens to be redeemed
+  /// @return The actual amount of tokens transferred
+  function _redeem(uint256 amount) internal override returns (uint256) {
```
Comment (Perseverance) [inline contracts/prize-pool/barnbridge/BarnPrizePool.sol]: I think this logic might be a bit too simple and is not accounting for early exits, but it all depends on the staking protocol behavior.
Lets say the lottery has 7 more days to go and someone stakes 1 BARN. On day 6 (1 day before the lottery end) this 1 BARN has accrued 0.1 BARN. It is very important that the user if the user decides to exit, he is given back maximum of 1 BARN and the 0.1 BARN accrued is returned to the staking protocol and eventually gotten by the winner. Can you check and see if this is the case?

**Review by Perseverance — CHANGES_REQUESTED** [review]

File: contracts/external/barnbridge/BarnInterface.sol (hunk 3 above)
Comment (sstrenev) [reply contracts/external/barnbridge/BarnInterface.sol]: @Perseverance As far as I understood, this is it: https://github.com/LimeChain/BarnBridge-Barn/blob/master/contracts/facets/BarnFacet.sol

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol (hunk 4 above)
Comment (sstrenev) [reply contracts/prize-pool/barnbridge/BarnPrizePool.sol]: Yes, I agree. Basically in the protocol - there are two possibilities - deposit $BOND without locking it (nothing is accrued in this case) and deposit $BOND and lock it (maximum duration 1 year, where accrued amount is linear to the time staked, with maximum multiplier x2). However, when $BOND is locked, it cannot be withdrawn until the time has passed, so basically from what I see, the user wouldn't be able to make an early exit.

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol (hunk 1 above)
Comment (sstrenev) [reply contracts/prize-pool/barnbridge/BarnPrizePool.sol]: Yes, think this is redundant. Will remove it.

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    uint256 postBalance = token.balanceOf(address(this));
+    uint256 amountWithdrawn = postBalance.sub(preBalance);
+    
+    return amountWithdrawn;
+  }
+
+  /// @dev Gets the underlying asset token used by the Yield Service
+  /// @return A reference to the interface of the underling asset token
+  function _token() internal override view returns (IERC20Upgradeable) {
+    return IERC20Upgradeable(barn.token());
```
Comment (sstrenev) [inline contracts/prize-pool/barnbridge/BarnPrizePool.sol]: @Perseverance Now I see this is not exactly right, as the BarnFacet does not expose a token() function, so the BOND token address should be set in the constructor of the BarnPrizePool maybe?

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    );
+    barn = _barn;
+
+    emit BarnPrizePoolInitialized(address(barn));
+  }
+
+  /// @dev Gets the balance of the underlying assets held by the Yield Service
+  /// @return The underlying balance of asset tokens
+  function _balance() internal override returns (uint256) {
+    uint256 balance = barn.balanceOf(address(this));
```
Comment (Perseverance) [inline contracts/prize-pool/barnbridge/BarnPrizePool.sol]: Does this include the deposit and interest too?

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    return balance;
+  }
+
+  /// @dev Allows a user to supply asset tokens in exchange for yield-bearing tokens
+  /// to be held in escrow by the Yield Service
+  function _supply(uint256 amount) internal override {
+    IERC20Upgradeable assetToken = _token();
+    assetToken.approve(address(barn), amount);
+    barn.depositAndLock(amount, (block.timestamp + maxTimelockDuration));
```
Comment (Perseverance) [inline contracts/prize-pool/barnbridge/BarnPrizePool.sol]: Is there a chance for the timelock to extend past the end of the campaign. Shouldnt be the timelock always be end of campaign - now?

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  /// @return The actual amount of tokens transferred
+  function _redeem(uint256 amount) internal override returns (uint256) {
+    require(barn.userLockedUntil(msg.sender) < block.timestamp, "BarnPrizePool/user-locked");
+    IERC20Upgradeable token = _token();
+
+    require(_balance() >= amount, "BarnPrizePool/insuff-liquidity");
+
+    uint256 preBalance = token.balanceOf(address(this));
+
+    barn.withdraw(amount);
```
Comment (Perseverance) [inline contracts/prize-pool/barnbridge/BarnPrizePool.sol]: Does withdraw give you what you've staked or also the amount of internet you've accrued?

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol (hunk 7 above)
Comment (Perseverance) [reply contracts/prize-pool/barnbridge/BarnPrizePool.sol]: I think we should also override `withdrawInstantlyFrom` and disable it until the end of the campaign due to locking.

File: test/BarnPrizePool.test.js [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+const { expect } = require('chai')
+const buidler = require('@nomiclabs/buidler')
+
+const toWei = ethers.utils.parseEther
+
+const debug = require('debug')('ptv3:yVaultPrizePool.test')
+
+let overrides = { gasLimit: 20000000 }
+
+describe('BarnPrizePool', function () {
```
Comment (Perseverance) [inline test/BarnPrizePool.test.js:22]: I would like to see the following tests:
1. A test that calls the depositTo method of the prizePool and checks the deposited balances
2. A test that calls withdraw after the campaign is over and gets you back your BOND
3. A test that showcases the award is sent to the winner
4. A test that showcases that the `balanceOf` method of the Prize pool returns the interest accrued too

**Review by Perseverance — CHANGES_REQUESTED** [review]

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol (hunk 8 above)
Comment (sstrenev) [reply contracts/prize-pool/barnbridge/BarnPrizePool.sol]: This gives you only what you've staked. It comes from BarnFacet. The amount you have accrued comes from the Rewards.sol contract and claim() function.

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol (hunk 7 above)
Comment (sstrenev) [reply contracts/prize-pool/barnbridge/BarnPrizePool.sol]: Okay, yes will change that!

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol (hunk 6 above)
Comment (Perseverance) [reply contracts/prize-pool/barnbridge/BarnPrizePool.sol]: This should be owed[us]+userCalculatedSomething

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol (hunk 8 above)
Comment (Perseverance) [reply contracts/prize-pool/barnbridge/BarnPrizePool.sol]: Override the `award` and do the math checking if there is any balance in this contract.

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol (hunk 5 above)
Comment (Perseverance) [reply contracts/prize-pool/barnbridge/BarnPrizePool.sol]: Pass both BOND and Rewards addresses as params

**Review by Perseverance — CHANGES_REQUESTED** [review]

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol (hunk 5 above)
Comment (sstrenev) [reply contracts/prize-pool/barnbridge/BarnPrizePool.sol]: this is done now

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol (hunk 6 above)
Comment (sstrenev) [reply contracts/prize-pool/barnbridge/BarnPrizePool.sol]: this now includes the deposit + owed

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol (hunk 7 above)
Comment (sstrenev) [reply contracts/prize-pool/barnbridge/BarnPrizePool.sol]: the latest discussed logic is in the _redeem method and the lock has been removed

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol (hunk 8 above)
Comment (sstrenev) [reply contracts/prize-pool/barnbridge/BarnPrizePool.sol]: added the latest logic as discussed

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+        uint256 balance = barn.balanceOf(address(this));
+        uint256 owed = rewards.owed(address(this));
+        return balance.add(owed);
+    }
+
+    /// @dev Allows a user to supply asset tokens in exchange for yield-bearing tokens
+    /// to be held in escrow by the Yield Service
+    function _supply(uint256 amount) internal override {
+        IERC20Upgradeable bondToken = _token();
+        bondToken.approve(address(barn), bondToken.totalSupply());
```
Comment (Perseverance) [inline contracts/prize-pool/barnbridge/BarnPrizePool.sol]: Should we approve just `amount` ?

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+        /// Check if there is anything to claim from the rewards
+        if (amountToClaim > 0) {
+            rewards.claim();
+        }
+
+        uint256 currentBalance = token.balanceOf(address(this));
+
+        /// If current bond balance is enough, deposit the difference back to Barn
+        if (currentBalance >= amount) {
```
Comment (Perseverance) [inline contracts/prize-pool/barnbridge/BarnPrizePool.sol]: I think this check should be without the `=` (just `<`)

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+        /// If current bond balance is not enough, try to withdraw the difference from Barn
+        if (currentBalance < amount) {
+            diff = amount.sub(currentBalance);
+            barn.withdraw(diff);
+        }
+
+        uint256 postBalance = token.balanceOf(address(this));
+
+        require(postBalance >= amount, "BarnPrizePool/insuff-liquidity");
```
Comment (Perseverance) [inline contracts/prize-pool/barnbridge/BarnPrizePool.sol]: I think the postBalance should be exactly equal to `amount`

**Review by Perseverance — CHANGES_REQUESTED** [review]

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol (hunk 12 above)
Comment (sstrenev) [reply contracts/prize-pool/barnbridge/BarnPrizePool.sol]: Fixed :)

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol (hunk 11 above)
Comment (sstrenev) [reply contracts/prize-pool/barnbridge/BarnPrizePool.sol]: Fixed as well

File: contracts/prize-pool/barnbridge/BarnPrizePool.sol (hunk 10 above)
Comment (sstrenev) [reply contracts/prize-pool/barnbridge/BarnPrizePool.sol]: yes I agree, it is set to amount now

**Review by Perseverance — APPROVED** [review]

**Pull request merged**